src/Storage/Repository/PgInstanceRepository.cs

```diff
@@ -396,6 +390,77 @@ public async Task<Instance> Update(Instance instance, List<string> updatePropert
+            parameters.AddWithValue("_presentationtexts", NpgsqlDbType.Jsonb, updateProperties.Contains(nameof(instance.PresentationTexts)) ? instance.PresentationTexts : DBNull.Value);
+            parameters.AddWithValue("_status", NpgsqlDbType.Jsonb, updateProperties.Contains(nameof(instance.Status)) ? CustomSerializer.Serialize(instance.Status, updateProperties) : DBNull.Value);
+            parameters.AddWithValue("_substatus", NpgsqlDbType.Jsonb, updateProperties.Contains(nameof(instance.Status.Substatus)) ? instance.Status.Substatus : DBNull.Value);
+            parameters.AddWithValue("_process", NpgsqlDbType.Jsonb, updateProperties.Contains(nameof(instance.Process)) ? instance.Process : DBNull.Value);
+            parameters.AddWithValue("_lastchanged", NpgsqlDbType.TimestampTz, instance.LastChanged ?? DateTime.UtcNow);
+            parameters.AddWithValue("_taskid", NpgsqlDbType.Text, instance.Process?.CurrentTask?.ElementId ?? (object)DBNull.Value);
+        }
+
+        /// <inheritdoc/>
+        public async Task<Instance> Update(Instance instance, List<string> updateProperties, List<InstanceEvent> events)
+        {
+            if (events.Count > 500) 
+            {
+                // Batching is implemented by simply appending multiple value clauses to the insert statement
+                // There is a limit to how many parameters PostgreSQL can accept (the count must fit within a 16 bit integer)
+                // so we limit the event count to some large number we never expect to hit.
+                throw new InvalidOperationException("Too many events to insert in one go");
+            }
+            else if (events.Count == 0)
+            {
+                return await Update(instance, updateProperties);
+            }
+
+            // Remove last decimal digit to make postgres TIMESTAMPTZ equal to json serialized DateTime
+            instance.LastChanged = instance.LastChanged != null ? new DateTime((((DateTime)instance.LastChanged).Ticks / 10) * 10, DateTimeKind.Utc) : null;
+            List<DataElement> dataElements = instance.Data;
+
+            ToInternal(instance);
+            instance.Data = null;
+            await using var batch = _dataSource.CreateBatch();
+
+            var updateCommand = new NpgsqlBatchCommand(_updateSql);
+            BuildUpdateCommand(instance, updateProperties, updateCommand.Parameters);
+            batch.BatchCommands.Add(updateCommand);
+
+            var insertEventsSql = new StringBuilder(_insertInstanceEventsSqlTemplate);
+
+            for (int i = 0; i < events.Count; i++)
+            {
+                insertEventsSql.Append($"{(i == 0 ? string.Empty : ", ")}(@_{i}instance, @_{i}alternateid, jsonb_strip_nulls(@_{i}event))");
+            }
+
+            var insertEventsComand = new NpgsqlBatchCommand(insertEventsSql.ToString());
```

Comment:
We've been trying to use explicit types. Less important in this case, but nice to have when it's not immediately clear what a method returns.
```suggestion
            NpgsqlBatchCommand insertEventsComand = new(insertEventsSql.ToString());
```